std/haxe/math/bigint/BigInt.hx

```diff
@@ -0,0 +1,450 @@
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package haxe.math.bigint;
+
+import haxe.ds.Vector;
+import haxe.io.Bytes;
+
+/* Original code courtesy Chuck Batson (github.com/cbatson) */
+@:allow(haxe.math.bigint)
+abstract BigInt(BigInt_) {
+	//-----------------------------------------------------------------------
+	// Public constants
+	//-----------------------------------------------------------------------
+	public static var ZERO(default, null):BigInt = new BigInt(BigInt_.fromInt(0));
+	public static var ONE(default, null):BigInt = new BigInt(BigInt_.fromInt(1));
+	public static var TWO(default, null):BigInt = new BigInt(BigInt_.fromInt(2));
+	public static var NEGATIVE_ONE(default, null):BigInt = new BigInt(BigInt_.fromInt(-1));
+
+	//-----------------------------------------------------------------------
+	// Public interface
+	//-----------------------------------------------------------------------
+
+	public inline function sign():Int {
+		return BigInt_.sign1(this);
+	}
+
+	public inline function isZero():Bool {
+		return BigInt_.isZero1(this);
+	}
+
+	public inline function isNegative():Bool {
+		return BigInt_.isNegative1(this);
+	}
+
+	public inline function isPositive():Bool {
+		return BigInt_.isPositive1(this);
+	}
+
+	public inline function isOdd():Bool {
+		return BigInt_.isOdd1(this);
+	}
+
+	public inline function isEven():Bool {
+		return BigInt_.isEven1(this);
+	}
+
+	public inline function min(other:BigInt):BigInt {
+		return new BigInt(this.min(other));
+	}
+
+	public inline function max(other:BigInt):BigInt {
+		return new BigInt(this.max(other));
+	}
+
+	public inline function toString():String {
+		return BigInt_.toString1(this);
+	}
+
+	public inline function toHex():String {
+		return BigInt_.toHex1(this);
+	}
+
+	public inline function toBytes():Bytes {
+		return BigInt_.toBytes1(this);
+	}
+
+	public inline function toInts(output:Vector<Int>):Int {
+		return BigInt_.toInts1(this, output);
+	}
+
+	public static inline function fromInt(value:Int):BigInt {
+		return new BigInt(BigInt_.fromInt(value));
```

Comment:
Maybe some caching could be added for values e.g.from -10 to 10, so that in these cases no new instances are created but shared BigInts are re-used. In Java for example values from -16  to 16 are cached https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/math/BigInteger.java#L949

Reply:
I'm not sure about that. What is the purpose of this caching in Java ? The only logic for that is if there are any operations and they often  use values between -16 and 16.  Otherwise , why  not cache between -100 and 100 or any other values ?

Reply:
It is about reducing garbage collector pressure from very commonly used values. why they decided for +/-16 I don't know. For the regular Java Integer type it is actually -128 to 128 https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/lang/Integer.java#L638

Maybe we can returned the cached values for the once static variables are initialized anyways? I.e. ZERO, ONE, TWO, NEGATIVE_ONE (which I would actually call MINUS_ONE - a symbol that is 25 times more common when you do a GitHub code search).

Reply:
Regarding the cache, since this is not my implementation, I just foundthat there is already a cache for values between -16 and 16 ( https://github.com/HaxeFoundation/haxe/blob/704100c4a8c57ed417dfc3c2bcf688dae77ca15b/std/haxe/math/bigint/BigInt_.hx#L334 ) , the fromInt() method  calls getCachedValue(...) ,so it caches all values ( -16...16) after first call . 
I think this is a good implementation and does not require change

 About NEGATIVE_ONE  , I changed it to MINUS_ONE